java/logic/RuleImpl.java

```diff
@@ -69,17 +71,35 @@ public ThingVariable<?> getThen() {
     @Override
     public void setLabel(TypeDBTransaction transaction, String newLabel) {
         if (newLabel == null || newLabel.isEmpty()) throw new TypeDBClientException(MISSING_LABEL);
-        rule_set_label(((LogicManagerImpl) transaction.logic()).nativeTransaction, nativeObject, newLabel);
+        try {
+            rule_set_label(nativeTransaction(transaction), nativeObject, newLabel);
+        } catch (com.vaticle.typedb.client.jni.Error e) {
+            throw new TypeDBClientException(e);
+        }
     }
 
     @Override
     public void delete(TypeDBTransaction transaction) {
-        rule_delete(((LogicManagerImpl) transaction.logic()).nativeTransaction, nativeObject);
+        try {
+            rule_delete(nativeTransaction(transaction), nativeObject);
+        } catch (com.vaticle.typedb.client.jni.Error e) {
+            throw new TypeDBClientException(e);
+        }
     }
 
     @Override
     public final boolean isDeleted(TypeDBTransaction transaction) {
-        return rule_is_deleted(((LogicManagerImpl) transaction.logic()).nativeTransaction, nativeObject);
+        try {
+            return rule_is_deleted(nativeTransaction(transaction), nativeObject);
+        } catch (com.vaticle.typedb.client.jni.Error e) {
+            throw new TypeDBClientException(e);
+        }
+    }
+
+    private static com.vaticle.typedb.client.jni.Transaction nativeTransaction(TypeDBTransaction transaction) {
+        com.vaticle.typedb.client.jni.Transaction nativeTransaction = ((LogicManagerImpl) transaction.logic()).nativeTransaction;
+        if (!nativeTransaction.isOwned()) throw new TypeDBClientException(TRANSACTION_CLOSED);
+        return nativeTransaction;
```

Comment:
In the new model we decided in Python, we can keep this method but have it get the `this.nativeObject()` with all the projection that entails. This means we keep the `transaction` as the nativeObject, and 'rename' it via a simple getter that redirects.